opt: factor check constraints into Scan FDs

RaduBerinde opened the pull request:
#### opt: build check constraint filters upfront

We use check constraint filters to try to improve index constraint generation. We
build all check constraints as scalars and store them in the TableMeta; then we
create the FiltersExpr in GenerateConstrainedScans.

With this change, we now create the filters upfront. This will allow other uses
for the filters (like better FD generation for Scans).

Release note: None

#### opt: factor check constraints into Scan FDs

The provided orderings logic has a fragility: if a child operator is "smarter"
about FDs than a parent operator, the parent operator might not be able to
calculate a provided ordering (the child ordering would seem incomplete). This
can happen when that child operator was created during exploration.

This change fixes an instance of this where the canonical scan doesn't take the
check constraints into account, but a constrained scan indirectly does (when the
checks are incorporated into the constraint). The fix is to take into account
the check constraints when calculating the Scan's FDs (which seems like a good
thing to do regardless of this issue).

Here's an example (simplified from #47041):
```
CREATE TABLE t (
  k INT8 PRIMARY KEY,
  a INT8 NOT NULL,
  b INT8 NOT NULL CHECK (b = 0),
  INDEX ba (b, a)
)
SELECT 1 FROM t WHERE a > 1 AND k > 0 GROUP BY b, a ORDER BY b
```

The optimized tree for this statement is a Project on top of DistinctOn on top
of a constrained Scan. DistinctOn requires `+b,+a` from its input (because it's
a streaming aggregation). The input provided ordering is just `+a` because the
constrained scan knows that `b` is constant. But the Select does not know that
`b` is constant, so DistinctOn does not realize that `+a` is not needed to
satisfy the required ordering. When we try to project away column `a`, we don't
know what to do with the `+a` ordering.

Fixes #47041.

Release note (bug fix): fixed an internal "no output column equivalent to X"
error in some very rare cases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47159)
<!-- Reviewable:end -->

andy-kimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-:-M4Mug7YBXftXLywgnmd:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @andy-kimball, @mgartner, and @rytaft)



<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 10 of 10 files at r1, 4 of 4 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-:-M4QwNWLAjD5TDmu4ywB:bg7igkv)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/table_meta.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4Qqq4S5sC4oT2K6ASt:-M4Qqq4S5sC4oT2K6ASu:bkz6d7i) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bb1b718d63943a68102d262338e6a900a35d022/pkg/sql/opt/table_meta.go#L135)):*
> ```Go
> 	IgnoreForeignKeys bool
> 
> 	// Constraints stores a *FiltersExpr containing filters that are known to
> ```

Is there any reason not to make the type `*FiltersExpr` then?
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 2054 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4Qw9EkCTquEjvDeW01:-M4Qw9EkCTquEjvDeW02:b-oq3hch) ([raw file](https://github.com/cockroachdb/cockroach/blob/359819dafda126c6d830081a167aab9fc88cf007/pkg/sql/opt/xform/testdata/physprops/ordering#L2054)):*
> ```
> ----
> 
> opt
> ```

[nit] are we hiding the ordering info from the output?


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47159)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/table_meta.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4Qqq4S5sC4oT2K6ASt:-M4R9hv48s2NWaS3cEug:b2dovzm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bb1b718d63943a68102d262338e6a900a35d022/pkg/sql/opt/table_meta.go#L135)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Is there any reason not to make the type `*FiltersExpr` then?
</blockquote></details>

Package deps :/
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 2054 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4Qw9EkCTquEjvDeW01:-M4R9pU5A9v8_p2F_td6:bhnbmhl) ([raw file](https://github.com/cockroachdb/cockroach/blob/359819dafda126c6d830081a167aab9fc88cf007/pkg/sql/opt/xform/testdata/physprops/ordering#L2054)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

[nit] are we hiding the ordering info from the output?
</blockquote></details>

No, there is no ordering here once the fact that b is constant is factored in.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47159)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/table_meta.go, line 135 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4Qqq4S5sC4oT2K6ASt:-M4RIfY2BAShY_xl3k8G:b-1vt1lw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bb1b718d63943a68102d262338e6a900a35d022/pkg/sql/opt/table_meta.go#L135)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Package deps :/
</blockquote></details>

Ah, right
___
*[pkg/sql/opt/xform/testdata/physprops/ordering, line 2054 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4Qw9EkCTquEjvDeW01:-M4RIoVv4XF87KECyO2m:b4c9lyr) ([raw file](https://github.com/cockroachdb/cockroach/blob/359819dafda126c6d830081a167aab9fc88cf007/pkg/sql/opt/xform/testdata/physprops/ordering#L2054)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

No, there is no ordering here once the fact that b is constant is factored in.
</blockquote></details>

:thumbsup: 


<!-- Sent from Reviewable.io -->

mgartner reviewed (approved):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47159)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @RaduBerinde)

___
*[pkg/sql/opt/optbuilder/select.go, line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4RQ8c66AJTcUQk0ajE:-M4RQ8c66AJTcUQk0ajF:bwc5tb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bb1b718d63943a68102d262338e6a900a35d022/pkg/sql/opt/optbuilder/select.go#L579)):*
> ```Go
> 
> 	var filters memo.FiltersExpr
> 	for ; chkIdx < numChecks; chkIdx++ {
> ```

Took me a few minutes to understand what you're doing here with `chkIdx` and above, but very cool. You're skipping ahead to the first validated constraint rather than starting at the first again. 

nit: add a quick comment here to explain that? If it's a common pattern we use then don't worry about it.


<!-- Sent from Reviewable.io -->

rytaft commented:
___
*[pkg/sql/opt/optbuilder/select.go, line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4RQ8c66AJTcUQk0ajE:-M4RajSl2bCczMcVs9FO:b-80dj41) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bb1b718d63943a68102d262338e6a900a35d022/pkg/sql/opt/optbuilder/select.go#L579)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Took me a few minutes to understand what you're doing here with `chkIdx` and above, but very cool. You're skipping ahead to the first validated constraint rather than starting at the first again. 

nit: add a quick comment here to explain that? If it's a common pattern we use then don't worry about it.
</blockquote></details>

+1 -- also took me a minute to see what was happening


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: build check constraint filters upfront

We use check constraint filters to try to improve index constraint generation.
We build all check constraints as scalars and store them in the TableMeta; then
we create the FiltersExpr in GenerateConstrainedScans.

With this change, we now create the filters upfront. This will allow other uses
for the filters (like better FD generation for Scans).

Release note: None

RaduBerinde committed:
opt: factor check constraints into Scan FDs

The provided orderings logic has a fragility: if a child operator is "smarter"
about FDs than a parent operator, the parent operator might not be able to
calculate a provided ordering (the child ordering would seem incomplete). This
can happen when that child operator was created during exploration.

This change fixes an instance of this where the canonical scan doesn't take the
check constraints into account, but a constrained scan indirectly does (when the
checks are incorporated into the constraint). The fix is to take into account
the check constraints when calculating the Scan's FDs (which seems like a good
thing to do regardless of this issue).

Here's an example (simplified from #47041):
```
CREATE TABLE t (
  k INT8 PRIMARY KEY,
  a INT8 NOT NULL,
  b INT8 NOT NULL CHECK (b = 0),
  INDEX ba (b, a)
)
SELECT 1 FROM t WHERE a > 1 AND k > 0 GROUP BY b, a ORDER BY b
```

The optimized tree for this statement is a Project on top of DistinctOn on top
of a constrained Scan. DistinctOn requires `+b,+a` from its input (because it's
a streaming aggregation). The input provided ordering is just `+a` because the
constrained scan knows that `b` is constant. But the Select does not know that
`b` is constant, so DistinctOn does not realize that `+a` is not needed to
satisfy the required ordering. When we try to project away column `a`, we don't
know what to do with the `+a` ordering.

Fixes #47041.

Release note (bug fix): fixed an internal "no output column equivalent to X"
error in some very rare cases.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47159)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[pkg/sql/opt/optbuilder/select.go, line 579 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47159#-M4RQ8c66AJTcUQk0ajE:-M4RbovG5s-24M5cOV7-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/7bb1b718d63943a68102d262338e6a900a35d022/pkg/sql/opt/optbuilder/select.go#L579)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

+1 -- also took me a minute to see what was happening
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
> Reviewed 4 of 4 files at r4.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47159)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+
